### PR TITLE
CORDA-1334: port enterprise statemachine

### PR DESCRIPTION
**Pull request by exFalso**
What is this
------------

This PR introduces a rewrite of the flow framework to use a pure state machine.

The reason for this change is two-fold:
1. Clarify the semantics of the flow framework.
2. Allow multithreading of flows in enterprise.

Initially I started off by having the statemachine live in a separate fiber, this way the bookkeeping around the flow would be separate from the fiber running the user-provided flow logic.
This was conceptually very simple, however because of the way we expose the database connection we cannot do this, otherwise on a quasar yield database connections would be "leaked" into waiting fibers, eventually causing saturation of the connection pool and a subsequent deadlock of the flow thread pool. (This turns out to be a general problem wrt resource management, we must make sure that any resources tied to a flow are released before yielding, otherwise we may leak these in waiting fibers)

This means that the state machine had to be "fused" with the fiber running the flow because we must commit the database transaction before we yield. This in turn means that we can't do the standard eventloop-style state update, rather we need to hook into the flow's logic at different API calls to process pending events. This is done in FlowStateMachineImpl, whose responsibility thus became running of the flow logic and "executing" of the state machine transitions.

An alternative solution I tried was to keep the state machine in a separate fiber called the worker, use a non-yielding push onto the worker's quasar queue and blocking *the executor thread* of the flow fiber until the worker finished processing. However this resulted in very tricky control flow and also relied on multithreading. In the end I decided that the "fusion" solution would be the most appropriate, but other suggestions that simplify the transitioning are very welcome.

The state machine revolves around a couple of main datastructures:

* `StateMachineState`: This is a (mostly) immutable datastructure capturing the state of the flow (the mutable parts are never persisted). Note that the persisted state (checkpoint) is a substructure of this, the rest of the state is in-memory (like unacknowledged messages or the live FlowLogic reference).
* `Event`: This captures an event to trigger a state machine transition. Each flow has a queue of Events.
* `Action`: This is a reified representation of side-effecting/IO actions to be executed as part of state machine transitions. Examples are database commits and message sends/acknowledgements.
* `FlowContinuation`: This specifies what should happen once the transition succeeds. This is an unfortunate necessity to control the transition hooks, a result of the fact we can't use a simple eventloop.
* `TransitionResult`: a wrapper of the new `StateMachineState`, a `List<Action>` to execute and the `FlowContinuation`.

The transitions then are pure functions
```
fun transition(state: StateMachineState, event: Event): TransitionResult
```
Which expresses a transition into `newState` *IF* the list of `actions` execute cleanly.

`TransitionResult`s are "interpreted" by `TransitionExecutor`s whose reponsibility is to execute the transition actions and handle failures while doing so. These executors may be intercepted to allow inspection of the state machine and transitions.

This rewrite allows easier reasoning about the flow and clarifies the semantics of some corner-case behaviours (which were misbehaving previously).

Behavioural changes
-------------------

Error handling: the concept of "flow erroredness" was introduced. When a flow is errored it still has a corresponding state machine to handle error propagation and possibly resumption, but control is not returned to the user-provided logic.

FlowHospital: an interface was introduced for the future hospital implementation, and a corresponding interceptor was hooked for notification of erroredness. The current implementation simply triggers error propagation immediately.

Session erroredness: session error conditions were separated from flow error conditions. Errors in sessions only trigger flow erroredness if the errored session is interacted with.

Error IDs: an opaque ID was introduced per "fresh" error to allow correlation of error conditions in different flows. This is propagated in rethrown exceptions.

Sessions: multiple sessions per Party were introduced and the old API was removed.

Checkpoints: the ID was changed from the hash of the checkpoint to `StateMachineRunId`

Deduplication: was fixed, and an explicit deterministic deduplication ID was introduced that bases the ID on a seed + number of state machine transitions

Batched IO requests: send/receive/getFlowInfo may be batched now (not exposed).

I've also included a bunch of additional features to provide more visibility and debuggability to the flow framework.

### CODE REVIEW

File: node/src/main/kotlin/net/corda/node/services/statemachine/FlowMessaging.kt [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        val address = serviceHub.networkService.getAddressOfParty(partyInfo)
+        val sequenceKey = when (message) {
+            is InitialSessionMessage -> message.initiatorSessionId
+            is ExistingSessionMessage -> message.recipientSessionId
+        }
+        serviceHub.networkService.send(networkMessage, address, sequenceKey = sequenceKey)
+    }
+
+    private fun SessionMessage.additionalHeaders(target: Party): Map<String, String> {
+
```
Comment (rick-r3) [inline node/src/main/kotlin/net/corda/node/services/statemachine/FlowMessaging.kt]: Remove blank line

File: node/src/main/kotlin/net/corda/node/services/statemachine/FlowMessaging.kt [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+     * Start the messaging using the [onMessage] message handler.
+     */
+    fun start(onMessage: (ReceivedMessage, deduplicationHandler: DeduplicationHandler) -> Unit)
+}
+
+/**
+ * Implementation of [FlowMessaging] using a [ServiceHubInternal] to do the messaging and routing.
+ */
+class FlowMessagingImpl(val serviceHub: ServiceHubInternal): FlowMessaging {
+
```
Comment (rick-r3) [inline node/src/main/kotlin/net/corda/node/services/statemachine/FlowMessaging.kt]: Remove blank line

File: core/src/main/kotlin/net/corda/core/flows/FlowException.kt [hunk 3]
```diff
@@ -25,6 +36,7 @@ open class FlowException(message: String?, cause: Throwable?) : CordaException(m
  * that we were not expecting), or the other side had an internal error, or the other side terminated when we
  * were waiting for a response.
  */
-class UnexpectedFlowEndException(message: String?, cause: Throwable?) : CordaRuntimeException(message, cause) {
-    constructor(msg: String) : this(msg, null)
-}
\ No newline at end of file
+class UnexpectedFlowEndException(message: String, cause: Throwable?, val originalErrorId: Long) :
```
[the remark this answers is not in the document]
Comment (exFalso) [reply core/src/main/kotlin/net/corda/core/flows/FlowException.kt]: I don't think it breaks (as we never serialise this), but there's no harm in adding back the old constructors

File: node/src/main/kotlin/net/corda/node/services/persistence/DBTransactionStorage.kt [hunk 4]
```diff
@@ -96,6 +100,17 @@ class DBTransactionStorage(cacheSizeBytes: Long) : WritableTransactionStorage, S
         }
     }
 
+    override fun trackTransaction(id: SecureHash): CordaFuture<SignedTransaction> {
+        return txStorage.locked {
+            val existingTransaction = get(id)
+            if (existingTransaction == null) {
+                updatesPublisher.filter { it.id == id }.toFuture()
```
[the remark this answers is not in the document]
Comment (exFalso) [reply node/src/main/kotlin/net/corda/node/services/persistence/DBTransactionStorage.kt]: Yes, this is a bug!

File: node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt [hunk 5]
```diff
@@ -29,32 +35,30 @@ class DBCheckpointStorage : CheckpointStorage {
             var checkpoint: ByteArray = EMPTY_BYTE_ARRAY
     ) : Serializable
 
-    override fun addCheckpoint(checkpoint: Checkpoint) {
-        currentDBSession().save(DBCheckpoint().apply {
-            checkpointId = checkpoint.id.toString()
-            this.checkpoint = checkpoint.serializedFiber.bytes // XXX: Is copying the byte array necessary?
+    override fun addCheckpoint(id: StateMachineRunId, checkpoint: SerializedBytes<Checkpoint>) {
+        currentDBSession().saveOrUpdate(DBCheckpoint().apply {
```
[the remark this answers is not in the document]
Comment (exFalso) [reply node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt]: Yes, this is used both for the first checkpoint and subsequent ones. BTW I had a branch somewhere that optimised this (to only `save` the first one and only `update` later ones) and it resulted in some speedup, so we may want to look into that

**Comment by exFalso** [comment]
@rick-r3 I cherry picked the soft lock release thing and addressed all other comments, back to you

File: node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (rick-r3) [reply node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt]: Yeah, that's why I was asking :-)

**Pull request merged**